src/Sylius/Bundle/CoreBundle/Doctrine/ORM/OrderRepository.php

```diff
@@ -431,4 +431,18 @@ public function findCartByTokenValue(string $tokenValue): ?BaseOrderInterface
             ->getOneOrNullResult()
         ;
     }
+
+    public function findCartByTokenValueAndChannel(string $tokenValue, ChannelInterface $channel): ?BaseOrderInterface
+    {
+        return $this->createQueryBuilder('o')
+            ->andWhere('o.state = :state')
+            ->andWhere('o.tokenValue = :tokenValue')
+            ->andWhere('o.channel = :channel')
+            ->setParameter('state', OrderInterface::STATE_CART)
+            ->setParameter('tokenValue', $tokenValue)
+            ->setParameter('channel', $channel)
+            ->getQuery()
+            ->getOneOrNullResult()
```

Comment:
Shouldn't we handle the `NonUniqueException` and return null ? Or do you think it is rather up to the calling script to do this ?
But in the second case, then we should add PHPDoc for the throw. Shouldn't we ?

Reply:
I think no ;) `tokenValue` is unique in DB so with the channel we are sure that we will get always a single cart or null :)
check this config: https://github.com/Sylius/Sylius/blob/master/src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/Order.orm.xml#L76

Reply:
`tokenValue` is unique, so IMO there is no need to use this exception